### ISSUE TITLE
FTL seems to handle recursion incorrectly

### ISSUE THREAD

**Issue by dever23b** [pi-hole/FTL#738]
**In raising this issue, I confirm the following (please check boxes, eg [X]) Failure to fill the template will close your issue:**

- [x] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md).
- [x] The issue I am reporting can be *replicated*
- [x] The issue I am reporting isn't a duplicate 

**How familiar are you with the codebase?:**

2

---

There is a bug open on the pi-hole/docker-pi-hole issue tracker here: https://github.com/pi-hole/docker-pi-hole/issues/600

The docker container reports as unhealthy and there are numerous log entries indicating that requests for pi.hole from 127.0.0.1 are being forwarded upstream.  We isolated the issue to the Health Check in the Dockerfile, specifically the +norecurse query used to confirm DNS resolution in the container.  In attempting to troubleshoot the issue, I believe the root of the issue is not the Docker health check, but rather how FTL handles non-recursive queries.  dig +recurse should be a valid test of whether Pi-hole DNS resolution is working.

**[BUG] Expected Behaviour:**
When FTL receives a query with norecurse set, it does not attempt to forward the query upstream and only answers locally.

**[BUG] Actual Behaviour:**
When FTL receives a query with norecurse set, it does not attempt to answer the query locally and forwards the query upstream, bypassing any applicable local response.

**[BUG] Steps to reproduce:**

- dig +norecurse +retry=0 @127.0.0.1 pi.hole

**Log file output [if available]**

```
dnsmasq[507]: query[A] pi.hole from 127.0.0.1
dnsmasq[507]: forwarded pi.hole to [upstream IP list]
```

**Device specifics**

Hardware Type: virtual
OS: Ubuntu 18.04 LTS / docker

**Additional Information**
Please see https://github.com/pi-hole/docker-pi-hole/issues/600 for more detail.  I believe that the issue boils down to the interaction between rfc1035.c:answer_request() and forward.c:receive_query() causing FTL to skip attempting local resolution before forwarding a query upstream.  I believe that this is contrary to the design of a non-recursive DNS query.

_This template was created based on the work of [`udemy-dl`](https://github.com/nishad/udemy-dl/blob/master/LICENSE)._

**Comment by DL6ER** [comment]
Quoting from the issue you referenced:
> I believe what may be happening is that, when `norecurse` is set, FTL doesn't even attempt to answer the query locally and instead forwards it.

I cannot reproduce this in a first attempt (note that I did not use a docker container!).
```
$ dig +norecurse +retry=0 @127.0.0.1 pi.hole

; <<>> DiG 9.11.3-1ubuntu1.11-Ubuntu <<>> +norecurse +retry=0 @127.0.0.1 pi.hole
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 8661
;; flags: qr aa ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;pi.hole.                       IN      A

;; ANSWER SECTION:
pi.hole.                2       IN      A       192.168.2.10

;; Query time: 0 msec
;; SERVER: 127.0.0.1#53(127.0.0.1)
;; WHEN: Mon Apr 27 20:09:57 UTC 2020
;; MSG SIZE  rcvd: 52
```
leading to:
```
Apr 27 20:09:57 dnsmasq[1375]: query[A] pi.hole from 127.0.0.1
Apr 27 20:09:57 dnsmasq[1375]: /etc/pihole/local.list pi.hole is 192.168.2.10
```

So the fact that this does not work seems to be related to your container missing `/etc/pihole/local.list`. Am I correct? If so, we should find out why this is the case.

Else than that, I can reproduce what you see using an actually non-existing domain such as
```
$ dig +norecurse +retry=0 @127.0.0.1 pi.hole5

; <<>> DiG 9.11.3-1ubuntu1.11-Ubuntu <<>> +norecurse +retry=0 @127.0.0.1 pi.hole5
; (1 server found)
;; global options: +cmd
;; connection timed out; no servers could be reached
```
```
Apr 27 20:14:44 dnsmasq[1375]: query[A] pi.hole5 from 127.0.0.1
Apr 27 20:14:44 dnsmasq[1375]: forwarded pi.hole5 to 127.0.0.1
```

I believe your source code investigation was close, however, this is not the most recent code any longer (you should look at the `development` branch for this). This last modification to this code was in https://github.com/pi-hole/FTL/commit/f519666e2abe1b8bbc957c776419fc431897fa40.

In fact, you are right in your assumption that
>  I believe that the issue boils down to the interaction between `rfc1035.c:answer_request()` and `forward.c:receive_query()` causing FTL to skip attempting local resolution before forwarding a query upstream.

The code in `forward.c:receive_query()` lack the ability for replying with `NXDOMAIN` (or, probably, an empty `NOERROR`) if `m == 0` - and this regardless of the state of the `rd` ("recursion desired") bit:
```c
      if (m >= 1)
        {
          send_from(listen->fd, option_bool(OPT_NOWILD) || option_bool(OPT_CLEVERBIND),
                    (char *)header, m, &source_addr, &dst_addr, if_index);
          daemon->metrics[METRIC_DNS_LOCAL_ANSWERED]++;
        }
      else if (!(header->hb3 & HB3_RD))
        {
          /* Insert me here: send NOERROR */
        }
      else if (forward_query(listen->fd, &source_addr, &dst_addr, if_index,
                             header, (size_t)n, now, NULL, ad_reqd, do_bit))
        daemon->metrics[METRIC_DNS_QUERIES_FORWARDED]++;
      else
        daemon->metrics[METRIC_DNS_LOCAL_ANSWERED]++;
```

> **[BUG] Expected Behaviour:**
> When FTL receives a query with norecurse set, it does not attempt to forward the query upstream and only answers locally.

You are right. When a DNS server receives a non-recursive request, it responds immediately with whatever local data it has available at the time without doing any additional processing. That may also mean: I have no answer for you.

You should file a bug to dnsmasq-discuss@lists.thekelleys.org.uk as we cannot implement bug fixes on our own to avoid incompatibilities with future updates from `dnsmasq`.

**Comment by dever23b** [comment]
I'm sorry, I should have clarified that point.  Local.list does exist in the container, and (provided I don't do a norecurse query) local.list is consulted for the response, and the logs indicate as much.  If I do dig without norecurse, the pihole.log indicates that it retrieved the answer from local.list.  When faced with a norecurse query, pihole does not appear to consult local.list.

**Comment by DL6ER** [comment]
> When faced with a norecurse query, pihole does not appear to consult local.list.

This is a bug which was closed in dnsmasq v2.81 which is included in the upcoming Pi-hole v5.0.

However, the issues you reported (NC queries should not be forwarded at all) still exists.

**Comment by dever23b** [comment]
Ok thanks.  Does this mean there's no need to file the additional bug at dnsmasq-discuss?

**Comment by DL6ER** [comment]
There is a bug:

> the issues you reported (NC queries should not be forwarded at all) still exists.

Even with the most recent dnsmasq v2.81.

**Comment by DL6ER** [comment]
@dever23b Did you write to the dnsmasq mailing list?

**Comment by dever23b** [comment]
This fell through the cracks.  I apologise.  I will try to get that filed this week.  Thanks for the reminder.

**Comment by dever23b** [comment]
@DL6ER I wasn't able to get this escalated to dnsmasq, but maybe someone else did.  I was working on my docker infrastructure and noticed that a v5.0 of Pi-Hole had been released.  After pulling the new image, I'm no longer seeing the same symptoms.  When I do a dig +norecurse, I'm getting the expected `/etc/pihole/local.list pi.hole is ...`

Can anyone else confirm whether they're seeing the same thing?  This may be resolved with FTL 5.0.

**Comment by DL6ER** [comment]
Pi-hole v5.0 contains the version 2.81 of `dnsmasq`, just released before. More than one year of development went into the new version and the amount of changes is quite a lot, so it may be difficult to track down what caused the improvement for you: https://github.com/pi-hole/FTL/compare/v4.3.1...v5.0

As long as this is solved for you, I will mark this as "Fixed in the next release" so it will be closed, soon. Thanks for coming back to us!

**Comment by DL6ER** [comment]
Closing as seems to be resolved with v5.0. Feel free to re-open of create a new issue ticket if this is not the case.
